### ISSUE TITLE
Autocomplete building

### ISSUE THREAD

**Issue by Valansch** [Valansch/RedMew#71]
Insert fuel/ammo in trains, tanks, cars, turrets etc. from the players inventory after they are placed.

**Comment by TWLTriston** [comment]
What kind of fuel? Wooden poles and chests? Should we use best fuel or worst fuel order ?

Ammo should we take from player gun if not enough?

**Comment by Valansch** [comment]
I think half a stack of the best fuel or ammo in the inventory would be a good idea. We should not take from the players ammo belt. Factorio handles the inventory seperately from ammo and tool belt to thats not a problem. I also think we should exclude poles and chests :D

**Comment by TWLTriston** [comment]
So Fuel order -- Rocket Fuel, Solid Fuel, Coal, Raw Wood
Ammo Order -- 
Uranium, Armor Piercing, Firearm
Flamethrower ammo
Explosive Uranium Cannon, Uranium Cannon shells, Explosive Cannon, Cannon shells

**Comment by plague006** [comment]
For reference:
https://github.com/3RaGaming/3Ra-Enhanced-Vanilla/blob/master/autofill.lua
https://github.com/3RaGaming/3Ra-Enhanced-Vanilla/blob/master/autofill_loader.lua

**Comment by grilledham** [comment]
I feel that auto-fill for turrets is too strong, which is why I haven't added this feature in the past. I have no objections to auto-fill for fuel, but it seems minor.
Thoughts on balance / whether auto-fill for turrets on some maps might be ok?

**Comment by plague006** [comment]
As quality of life, having it for locomotives would be amazing. 

I personally dislike it for fuel in furnaces and the like since it steals the wood from your inventory in the early gane. (And realistically furnaces and boilers always get fueled automatically)

Auto-fill for turrets is nice in SP. It's nice for placing defences quickly. But when used offensively it's very OP. We *could* check biters in range of the placed turret, but I think communicating that functionality to players would ultimately prove too big a problem. So in the end, I would opt to just not add that feature in the first place.

Cars, tanks, and trains being auto-filled wouldn't be great. Nothing else IMO.

**Comment by grilledham** [comment]
I'm confused your first sentence says having auto-fill for locomotives would be amazing, then you your last says 'Cars, tanks, and **trains** being auto-filled wouldn't be great.'

**Comment by iltar** [comment]
Is this still something we'd like to implement? Sounds like only locomotives would benefit from this greatly. Would this feature be worth quickly inserting a stack of fuel? It sounds more complex than worth it. What fuel to take? Wood? Coal? Nuclear? What about wooden boards? Power poles?

**Comment by plague006** [comment]
It's something that would be nice to have but it's such a low priority that
it's doubtful it'll ever get done.

**Comment by Valansch** [comment]
I dont think its too complex, its pretty simple. It's just that nobody really has gotten around to doing it :P
Its not just trains but also furnaces.

**Comment by mheguy** [comment]
I've actually always hated when furnaces get auto-filled since it takes your wood early-game. So I think if we to furnaces we need a toggle for the whole thing.

**Comment by linaori** [comment]
Perhaps we can create a user setting GUI to enable/disable these things? Would work cross-server with grilledham's new server system

**Comment by theorangeangle** [comment]
Aside from locomotives this feels a bit cheaty IMO - maybe as a one time perk you can buy from market?

**Comment by mheguy** [comment]
Is anyone interested in taking on the implementation of this feature? If nobody wants to assign this to themselves I will close this under “will not implement”.

**Comment by SimonFlapse** [comment]
If it's truly wanted, I could implement it. But I'd say I'm not sure about the usages. It's a minor QoL improvement for spawning vehicles with fuel already inserted, but the rest I don't really feel the same about. 

If I where to make it I'd do the following for trains, cars and tanks:
Find the players best available fuel -> Insert up-to one stack (Maybe two for nuclear fuel)

For furnaces we couldn't insert a full stack since you'd normally be laying down a ton of furnaces at a time. The player would simply run out of fuel multiple times. The solution could be to only insert 1/5 of a stack.

Implementation wise I'd skip all wood fuels since there's quite a bit.
Coal -> Solid Fuel -> Rocket Fuel -> Nuclear Fuel

If autofill for turrets where to happen I'd just limit it to 10 magazines. This would give the player enough time to refill it, without draining their inventory. They could always do CTRL + Right Click to will it afterwards if they have plenty of ammo.

I'd not add any gui to control it, only a command to toggle it.

We could lock it behind a research like automation 2 or buying from the market

**Comment by mheguy** [comment]
I personally only see the train as being both practical and not breaking balance. I could go along with the car and tank as the gameplay impact is negligible.

I personally find furnace auto-filling annoying since fuel to them is automated. Turret auto-filling wrecks balance.

**Comment by Valansch** [comment]
I don't think it breaks balance. Yes it makes it easier, but we are using a lot of tools to make the game harder aswell. For me auto insert into turrets just removes some frustration from the game. I could do the same with a macro. If you find it makes the game to easy just buff the biters :P

**Comment by Valansch** [comment]
I don't think it breaks balance. Yes it makes it easier, but we are using a lot of tools to make the game harder aswell. For me auto insert into turrets just removes some frustration from the game. I could do the same with a macro. If you find it makes the game to easy just buff the biters :P

**Comment by theorangeangle** [comment]
having seen a streamer play with an auto-fill mod on turrets in crash site, I would say it DEFINITELY has an effect on balance.

**Comment by grilledham** [comment]
#952 added autofill for turrets, with the trade off that turrets take time before they become activated.

I don't know if we want to add more types of atuofill, like fuel for vehicles / furnaces?

**Comment by linaori** [comment]
That would be a nice addition imo!

**Comment by theorangeangle** [comment]
#952 added autofill for turret ammo, with a delay. Do we want to add autofill for fuel as well?
